app/assets/javascripts/discourse/app/form-kit/components/fk/form.gjs

```diff
@@ -0,0 +1,340 @@
+
+  constructor() {
+    super(...arguments);
+
+    this.args.onRegisterApi?.({
+      set: this.set,
+      submit: this.onSubmit,
+      reset: this.onReset,
+    });
+
+    this.router.on("routeWillChange", this.checkIsDirty);
+  }
+
+  willDestroy() {
+    super.willDestroy();
+
+    this.router.off("routeWillChange", this.checkIsDirty);
+  }
+
+  get mutable() {
+    return this.args.mutable ?? false;
+  }
+
+  @action
+  checkIsDirty(transition) {
+    if (
+      this.isDirtyForm &&
+      !transition.isAborted &&
+      !transition.queryParamsOnly
+    ) {
+      transition.abort();
+
+      this.dialog.yesNoConfirm({
+        message: I18n.t("form_kit.dirty_form"),
+        didConfirm: () => {
+          this.isDirtyForm = false;
+          this.onReset();
+          transition.retry();
+        },
+      });
+    }
+  }
+
+  @cached
+  get effectiveData() {
+    const obj = this.args.data ?? {};
+
+    if (this.mutable) {
+      return obj;
+    }
+
+    const { transientData } = this;
+
+    return new Proxy(obj, {
+      get(target, prop) {
+        return prop in transientData
+          ? transientData[prop]
+          : Reflect.get(target, prop);
+      },
+
+      set(target, property, value) {
+        return Reflect.set(transientData, property, value);
+      },
+
+      has(target, prop) {
+        return prop in transientData ? true : Reflect.has(target, prop);
+      },
+
+      getOwnPropertyDescriptor(target, prop) {
+        return Reflect.getOwnPropertyDescriptor(
+          prop in transientData ? transientData : target,
+          prop
+        );
+      },
+
+      ownKeys(target) {
+        return (
+          [...Reflect.ownKeys(target), ...Reflect.ownKeys(transientData)]
+            // return only unique values
+            .filter((value, index, array) => array.indexOf(value) === index)
+        );
+      },
+
+      deleteProperty(target, prop) {
+        if (prop in transientData) {
+          delete transientData[prop];
+        }
+
+        return true;
+      },
+    });
+  }
+
+  get validateOn() {
+    return this.args.validateOn ?? VALIDATION_TYPES.submit;
+  }
+
+  get fieldValidationEvent() {
+    const { validateOn } = this;
+
+    if (validateOn === VALIDATION_TYPES.submit) {
+      return undefined;
+    }
+
+    return validateOn;
+  }
+
+  get hasErrors() {
+    return Array.from(this.fields.values()).some((field) => field.hasErrors);
+  }
+
+  get errors() {
+    const errors = {};
+    for (const [name, field] of this.fields) {
+      errors[name] = field.errors;
+    }
+    return errors;
+  }
+
+  @action
+  addError(name, message) {
+    const field = this.fields.get(name);
+    field?.pushError(message);
+  }
+
+  @action
+  async set(key, value) {
+    this.isDirtyForm = true;
+    set(this.effectiveData, key, value);
+
+    if (this.fieldValidationEvent === VALIDATION_TYPES.change) {
+      await this.handleFieldValidation(key);
+    }
+
+    await new Promise((resolve) => next(resolve));
```

Comment:
I would like to avoid this, but couldn't figure out how to without having to have many `settled()` in my tests